### PR TITLE
receiving_key_scopes migration: Fall back to note commitment matching for scope determination

### LINKED ISSUE #1117: `zcash_client_sqlite`: `receiving_key_scopes` migration appears broken for existing wallets
I have a local test wallet DB working with revision 236cd569ee4a824d98920d1a61b8cc2b90ceba1d. I tried to upgrade it to latest `main` and got the following error:

```
2024-01-17T03:08:09.323554Z  INFO schemer: Migrating up to target: None    
2024-01-17T03:08:09.323897Z  INFO schemer: Migrating up to target: None    
2024-01-17T03:08:09.323928Z  INFO schemer: Applying migration ee89ed2b-c1c2-421e-9e98-c1e3e54a7fc2    
Error: An error occurred while applying migration ee89ed2b-c1c2-421e-9e98-c1e3e54a7fc2 (Add decryption key scope to persisted information about received notes.) up: Invalid column type Null at index: 2, name: raw.

Caused by:
    0: Invalid column type Null at index: 2, name: raw
    1: Invalid column type Null at index: 2, name: raw
```

The migration in question was added in #1055, and AFAICT no changes were made to existing migrations since 236cd569ee4a824d98920d1a61b8cc2b90ceba1d, so AFAICT this migration is intended to work (but of course could itself still change before the next crate release of `zcash_client_sqlite`).

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #1117

### CODE REVIEW

**Comment by str4d** [comment]
This migration fails with a different error:
```
2024-01-24T18:12:08.307240Z  INFO schemer: Applying migration ee89ed2b-c1c2-421e-9e98-c1e3e54a7fc2    
Error: An error occurred while applying migration ee89ed2b-c1c2-421e-9e98-c1e3e54a7fc2 (Add decryption key scope to persisted information about received notes.) up: Invalid column type Null at index: 4, name: expiry_height.

Caused by:
    0: Invalid column type Null at index: 4, name: expiry_height
    1: Invalid column type Null at index: 4, name: expiry_height
```

File: zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs [hunk 1]
```diff
@@ -111,13 +177,85 @@ impl<P: consensus::Parameters> RusqliteMigration for Migration<P> {
...
+                let scope = select_note_scope(
+                    &mut commitment_tree,
+                    dfvk,
+                    &diversifier,
+                    &note_value.try_into().unwrap(),
+                    &rseed,
+                    note_commitment_tree_position,
                 )?;
+
+                if scope == Some(Scope::Internal) {
```
Comment (nuttycom) [inline zcash_client_sqlite/src/wallet/init/migrations/receiving_key_scopes.rs:252]: In the case where `scope == None`, it means that we have no way to spend the note, so the scope of the key is irrelevant. We _could_ add a sentinel for this, but I think that just leaving it as `External` is probably okay?

**Review by str4d — APPROVED** [review]
ACK 55a8668a3b45306e580114f2bf313ee891e58e7e with non-blocking nits.

**Comment by str4d** [comment]
[Force-pushed](https://github.com/zcash/librustzcash/compare/55a8668a3b45306e580114f2bf313ee891e58e7e..d0a27eda90ffbc9cdf5367aa782a91522eb7f030) to resolve my comments and fix clippy lints.

**Review by str4d — APPROVED** [review]
Re-ACK d0a27eda90ffbc9cdf5367aa782a91522eb7f030

**Pull request merged**